### PR TITLE
[KERNEL-ESTIMATION] [ObjectivityLtd] [02.11.2022]

### PR DESCRIPTION
**Pull request by mbaczyk123**
### Team members
Michal Baczyk
Peter den Haan

### Submission summary
Accuracy: 100%.
HAQS-2022/qml-classifier-challenge/kernel_estimator_Michal.ipynb
Private submission. qBraid e-mail: mbaczyk@objectivity.co.uk

### Additional comments
I am working on the fork from [ObjectivityLtd:ObjectivityTeam] in the qBraid lab registed for the mbaczyk@objectivity.co.uk.

### CODE REVIEW

**Review by ryanhill1 — CHANGES_REQUESTED** [review]
Thanks for your submission! 

Your solution does not seem to have addressed [my comment from PR #30](https://github.com/qBraid/HAQS-2022/pull/30#pullrequestreview-1163592034). I'm not seeing where you define and optimize against the loss function $L_D$ in step 17 of Algorithm 1. It still seems as though you are just using `SVC.fit` with the defaults instead of fitting against the desired loss function. Additionally, your `classifier()` function still seems not to account for the $y_i$ term in the sum, as noted in the previous comment. Please let me know if I am misinterpreting any parts of your code.

Secondly, the first line of your `fidelity_estimate()` function causes infinite recursion:
```
def fidelity_estimate(x, z):
    """Directly estimate the fidelity between a pair of feature vectors for data `x` and `z`

    Args:
        x : shape (2,) tensor containing one input data vector
        z : shape (2,) tensor containing one input data vector

    Returns:
        shape (2, `R`) tensor containing `R` Z-basis measurement samples on each qubit, where `R` 
        is the number of shots determined from the dev.shots attribute of the corresponding device.
    """
    samples = fidelity_circuit(x,z)
   ...
```
so the program never actually finishes.

**Comment by mbaczyk123** [comment]
The way the SVC.fit function is currently used is mathematically exactly equivalent to the concave optimization problem given. 
See point 1.4.7.1. SVC "The dual problem" https://scikit-learn.org/stable/modules/svm.html#svm-kernels .
As you can see there the SVC is minimizing exactly the same expression just with the minus sign which is responsible from the switch from maximization to minimization. The SVC has also a small tweak to it since the alphas are bounded by C which might cause some differences between these two approaches but it is a minor detail not playing a role there. 

Additionally, if you look at the documentation of SVC, you see that dual_coef_ is exactly the alpha and y terms combined together. https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC
Hence, we are taking these into account. 

fidelity_estimate does not give infinite recursion. It uses fidelity_circuit function inside. 

The notebook runs smoothly for us and gives us 100% accuracy.

**Comment by mbaczyk123** [comment]
@ryanhill1 Please see my comment above and I would very much appreciate if you re-evaluate our submission.

**Comment by mbaczyk123** [comment]
@ryanhill1  I just reran the code and everything worked smoothly and we obtained 100% accuracy.

**Comment by ryanhill1** [comment]
> The way the SVC.fit function is currently used is mathematically exactly equivalent to the concave optimization problem given. See point 1.4.7.1. SVC "The dual problem" https://scikit-learn.org/stable/modules/svm.html#svm-kernels . As you can see there the SVC is minimizing exactly the same expression just with the minus sign which is responsible from the switch from maximization to minimization. The SVC has also a small tweak to it since the alphas are bounded by C which might cause some differences between these two approaches but it is a minor detail not playing a role there.
> 
> Additionally, if you look at the documentation of SVC, you see that dual_coef_ is exactly the alpha and y terms combined together. https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC Hence, we are taking these into account.
> 
> fidelity_estimate does not give infinite recursion. It uses fidelity_circuit function inside.
> 
> The notebook runs smoothly for us and gives us 100% accuracy.

Hi Michal, I was mistaken, there is no infinite recursion. I misread the function there. Thank you for providing these links that help map the `SVC.fit` functionality to the minimization problem at hand. With this clarification, I will accept your submission, and record it on the leaderboard at 100% accuracy. Your team is now eligible to be considered for the overall QML challenge winner.

For your final submission, I would highly encourage your team to write out in markdown exactly how the `SVC.fit` method maps to optimizing the $\alpha$'s in the loss function $L_D$. I'm now aware of the correlation, but I will only be one of a number of judges who will be independently reviewing the final submissions.
